scalyr_agent/builtin_monitors/kubernetes_events_monitor.py

```diff
@@ -510,13 +510,31 @@ def _check_nodes_for_leader(self, k8s, query_fields=""):
         @param k8s: a KubernetesApi object for querying the k8s api
         @query_fields - optional query string appended to the node endpoint to allow for filtering
         """
+        # Hard guard-rails to prevent the agent from consuming large amounts of memory.
+        if query_fields:
```

Comment:
This sounds good as the first step.

Next step would be (as discussed) implementing pagination where we only retrieve up to X nodes at once (and only limit sets of of fields / metadata field we need for leader election) and process this data incrementally to avoid memory overhead.

Ideally we would also use streaming JSON parser (so we parse result incrementally), but this may be hard to do with the requests abstraction which is used right now.

On a related note, we should make this limit a configuration option in case it ever needs to be changed.

And ideally we would also do some analysis - if we retrieve metadata + id + other mandatory fields for X nodes, what kind of temporary memory usage increase this represents so we can come up with the approximate limit based on this data.

Reply:
I guess technically a simpler solution would be to only include those X retrieved nodes in leader election process instead of all nodes.

That's probably sufficient in case this X (limit) is large enough.

Only thing I would prefer to do in case we go with this approach is to try to retrieve those X nodes at random, but that may not be possible with available pagination query parameters (so the election is not biased against first X nodes or similar).